### PR TITLE
Make sure inactive request generator is stopped before spawning new one in Pub / Sub consumer.

### PR DESCRIPTION
**Pull request by dhermes**
~~This makes the `request_queue` used by `Consumer._request_generator_thread()` local to the caller. This way on recovery a new queue can be used for the new thread.~~

### CODE REVIEW

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 1]
```diff
@@ -244,7 +252,19 @@ def _blocking_consume(self):
                 break
             except Exception as exc:
                 recover = self._policy.on_exception(exc)
-                if not recover:
+                if recover:
+                    with threading.Lock():
```
Comment (lukesneeringer) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: I am fairly certain that `with threading.Lock():` used in this way does not work, because it is always using a different lock. You need to save a single lock object to the consumer instance so you are calling the `__enter__` method of the same lock.

Basically, `self._recovery_lock = threading.Lock()` in the `__init__.py` file, and then `with self._recovery_lock:` here.

File: pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py [hunk 2]
```diff
@@ -60,12 +60,13 @@ def __init__(self):
     def __contains__(self, needle):
         return needle in self._helper_threads
 
-    def start(self, name, queue, target):
+    def start(self, name, queue_put, target):
```
Comment (lukesneeringer) [inline pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py]: What is the rationale behind this change?

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 3]
```diff
@@ -244,7 +252,19 @@ def _blocking_consume(self):
...
+                        # Must lock so that ``self.send_request()`` cannot add
+                        # more items to the queue while it is being replaced.
+                        previous_queue = self._request_queue
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: @lukesneeringer After chatting with @jonparrott it seems there is another option here.

If we could guarantee that `self._request_queue.empty()` is `True`, then we could just send `STOP` into the queue and wait until the `STOP` gets consumed, then we don't have to swap the queue.

(As you can see from the coverage failure, the queue will always be empty.)

Comment (lukesneeringer) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: That makes sense to me.

Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: #4504 is an implementation of that other option

File: pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py (hunk 2 above)
Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py]: See where I am sending in `Consumer.send_request()` rather than passing in `Consumer._request_queue`. The problem is that if you send in a queue and then swap it off of the `Consumer`, then the helper thread won't be able to talk to the active request generator.

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py (hunk 1 above)
Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: What is the issue with always using a different lock?

File: pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py (hunk 2 above)
Comment (lukesneeringer) [reply pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py]: Understood.

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py (hunk 1 above)
Comment (lukesneeringer) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: It will never actually lock anything.

Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: BLERG

Comment (lukesneeringer) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: Longer explanation: `threading.Lock()` creates a single, discrete lock. To illustrate, imagine a world where we have two separate locks that lock two separate things:

To illustrate, consider a kitchen with two cooks. Both cooks want to make scrambled eggs on the stove top, and they want to put bread in the toaster. Since there is only one stove top and only one toaster, the cooks can not use them at the same time, but it would be fine for one cook to use the stove top while the other one uses the toaster.

These are different lock instances:

```python
stove_lock = threading.Lock()
toaster_lock = threading.Lock()

with stove_lock:
    scramble_eggs()

with toaster_lock:
    toast_bread()
```

When you call `with toaster_lock:`, if another thread is currently executing code within a separate `with toaster_lock:`, then your thread will block until the other thread calls `toaster_lock.__exit__`. However, it does not matter if another thread is currently using `stove_lock`.

But when you just do `with threading.Lock():` then you are creating a _different_ lock instance every time, and it is not actually possible for anything else to be checking against the same lock. The other threads are just blindly using their own, distinct locks and not checking in with yours.

**Comment by dhermes** [comment]
@jonparrott @lukesneeringer PTAL

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 4]
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
...
+            ValueError: If ``.close()`` fails for any reason other than
+                "generator already executing".
+            ValueError: If the queue is not empty and the generator is
+                running.
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
+            except ValueError as exc:
+                if exc.args != ('generator already executing',):
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: This is not so great to have this be "stringly-typed", I could just assume that a `ValueError` means it's already executing.

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 5]
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
...
+
+        Raises:
+            ValueError: If ``.close()`` fails for any reason other than
+                "generator already executing".
+            ValueError: If the queue is not empty and the generator is
+                running.
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: There **might** is a seemingly impossible-to-trigger corner case where the generator is **still running** for the few nanoseconds **after** the blocking `.get()` returns. In that case, we'd either need to

- Somehow force the gRPC thread that is consuming the request generator to knock it off
- Just try to `.close()` one more time if we detect the queue isn't empty

----

FWIW I don't think it's worth accounting for that impossible-to-trigger corner case.

File: pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py (hunk 2 above)
Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py]: @lukesneeringer The rationale has change slightly, since I no longer swap out the request queue. However, due to the `_put_lock` in `send_request()`, it still makes sense to avoid passing the queue itself.

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py (hunk 4 above)
Comment (lukesneeringer) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: I trust your judgment on this.

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 6]
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
...
+                    raise
+                if not self._request_queue.empty():
+                    raise ValueError('Queue expected to be empty.')
+                # If we **cannot** close the request generator,
+                # then there is no blocking get on the queue. Since
+                # we have locked ``.put()`` this means that the
+                # queue **was** and remains empty.
+                self._request_queue.put(_helper_threads.STOP)
+                # Wait for the request generator to ``.get()`` the ``STOP``.
+                while not self._request_queue.empty():
```
Comment (jonparrott) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: if we use `task_done` we can use [queue.join](https://docs.python.org/2/library/queue.html#Queue.Queue.join).

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 7]
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
...
+            ValueError: If the queue is not empty and the generator is
+                running.
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
+            except ValueError as exc:
+                if exc.args != ('generator already executing',):
+                    raise
+                if not self._request_queue.empty():
```
Comment (jonparrott) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: I thought we discussed that it's perfectly valid for the queue to have some pending items in it?

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 8]
```diff
@@ -218,6 +223,54 @@ def _request_generator_thread(self):
...
+                # then there is no blocking get on the queue. Since
+                # we have locked ``.put()`` this means that the
+                # queue **was** and remains empty.
+                self._request_queue.put(_helper_threads.STOP)
+                # Wait for the request generator to ``.get()`` the ``STOP``.
+                _LOGGER.debug(
+                    'Waiting for active request generator to receive STOP')
+                while not self._request_queue.empty():
+                    pass
+                request_generator.close()
```
Comment (dhermes) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: Luckily I re-factored the unit tests, because it made me realize that this could fail, because the generator still has to do:

```python
if request == _helper_threads.STOP:
    logger.debug()
    break
```

after the blocking `get()` completes, so this call to `close()` might happen **before** the generator gets those instructions.

I'm inclined to say we should either:

- Drop this statement completely (we are "safe" since we know that `STOP` has been pulled off the queue)
- Do a `while True` loop that tries to `close()` and `break`-s if it succeeds
- Do a `while request_generator.gi_running: pass` loop to block until it stops running and then call `close()`

Comment (jonparrott) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: I am in favor of dropping `close()` as we know that as long as `STOP` (or *any* item in the queue) has been consumed post-RPC-cancellation that gRPC will exit the thread.

Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: Me too. Jinx.

**Comment by dhermes** [comment]
@lukesneeringer @jonparrott PTAL. Phew I think it's finally merge-able.

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py [hunk 9]
```diff
@@ -218,6 +223,71 @@ def _request_generator_thread(self):
...
+                    # ``get()`` and the next ``yield`` / ``break``. However,
+                    # for practical purposes, we only need to stop the request
+                    # generator if the connection has timed out due to
+                    # inactivity, which indicates an empty queue.
+                    _LOGGER.debug(
+                        'Request generator could not be closed but '
+                        'request queue is not empty.')
+                    return False
+                # If we **cannot** close the request generator,
+                # then there is no blocking get on the queue. Since
```
Comment (jonparrott) [inline pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: Do you mean that there *is* a blocking get?

File: pubsub/tests/unit/pubsub_v1/subscriber/test_consumer.py [hunk 10]
```diff
@@ -142,6 +140,106 @@ def test_start_consuming():
...
+
+
+def basic_queue_generator(queue, received):
+    while True:
+        value = queue.get()
+        received.put(value)
+        yield value
+
+
+def test_stop_request_generator_not_running():
```
Comment (jonparrott) [inline pubsub/tests/unit/pubsub_v1/subscriber/test_consumer.py]: I would love some comments explaining the scenarios that each test exercise.

File: pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py (hunk 9 above)
Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: Yes, thanks.

Comment (dhermes) [reply pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py]: Updated

**Comment by dhermes** [comment]
I added explanations for each unit test @jonparrott. Also updated one of them to use a real generator instead of a `mock`. Will merge once CI is green.

This code is no fun.
